design/encryption-data-at-rest.md

```diff
@@ -108,55 +110,41 @@ Below diagram depicts the end-to-end encryption workflow detailing various modul
-    }
-```
 
-Few benefits of the above proposed schemes are:
-* JSON input format is extensible (adding new fields is backward compatible).
+Given encryption keys will be needed as part of cluster-recovery, this process/role needs to be recruited at the start of the cluster-recovery process (just after the “master/sequencer” process/role recruitment). All other FDB processes will interact with this process to obtain encryption keys needed to encrypt and/or decrypt the data payload. 
 
-* Popular Cloud KMS “getPublicKey” API accepts “keyId” as a string, hence, API should be easy to integrate.
+`Note`: An alternative would be to incorporate the functionality into the ClusterController process itself, however, having clear responsibility separation would make design more flexible and extensible in future if needed.
 
-    1. AWS: https://docs.aws.amazon.com/cli/latest/reference/kms/get-public-key.html
-    2. GCP: https://cloud.google.com/kms/docs/retrieve-public-key   
+### KMS-Connector
 
-`Future improvements`: FDBMonitor at present will launch one KMS-Connector process per FDB cluster host. Though multiple KMS-Connector processes are launched, only one process (collocated with EncryptKeyServer) would consume cluster resources. In future, possible enhancements could be: 
+Implements a native FDB KMS framework allowing multiple interfaces to co-existing and enabling FDB <-> KMS communication. Salient features:
 
-* Enable FDBMonitor to launch “N” (configurable) processes per cluster.
-* Enable the FDB cluster to manage external processes as well.
+* Abstract `KmsConnector` class, the class enables a specilization implementation to implement `actor` supporting desired communication protocol.
+* `KmsConnectorInterface` defines the supported endpoints allowing EncryptKeyProxy to fetch/refresh encryption keys.
+* `--kms-connector-type` configuration parameter supplied via `foundationdb.conf` controls the runtime selection of KmsConnector selection.
 
-### Encrypt KeyServer
+### **RESTKmsConnector**
 
-Salient features include:
+Implements REST protocol communication support to interact with external KMS.
 
-* New FDB role/process to allow fetching of encryption keys from external KeyManagementService interfaces. The process connects to the KMS-Connector REST interface to fetch desired encryption keys.
+The `foundationdb.conf` needs to be updated to supply following configuration parameters:
 
-* On an encryption-key fetch from KMS-Connector, it applies HMAC derivative function to generate a new encryption key and cache it in-memory. The in-memory cache is used to serve encryption key fetch requests from other FDB processes. 
+* `--discover-kms-conn-url-file`: local filesystem file-path defining the URL to connect with KMS on startup. FDB support periodic refreshes of KMS URLs if supported.
+* `--kms-conn-validation-token-details`: local file system file(s) detailing the validation tokens needed by KMS to authorize FDB <-> GS communication.
```

Comment:
Do we document anywhere what the format of these validation tokes should be?

Reply:
There is a detailed documentation explaining the format of the tokens and to construct the token configuration string.

```
	// Details to fetch validation token from a localhost file
	// acceptable format: "<token_name1>$<absolute_file_path1>,<token_name2>$<absolute_file_path2>,.."
	// NOTE: 'token-name" can NOT contain '$' character
	init( REST_KMS_CONNECTOR_VALIDATION_TOKEN_DETAILS,             "");
```